// @irys/upload-starknet-node/dist/cjs/walletConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KnownAccountContracts = void 0;
const hexStringPostProcessor = (result) => {
    const result2 = result[0];
    if (!(typeof result2 === 'string'))
        throw new Error(`Incorrect result ${result2}, expected hex string`);
    return Buffer.from(result2.startsWith('0x') ? result2.slice(2) : result2, 'hex');
};
exports.KnownAccountContracts = new Map([
    [
        1,
        {
            name: 'Argent',
            abi: [
                {
                    name: 'argent::account::interface::IArgentUserAccount',
                    type: 'interface',
                    items: [
                        {
                            name: 'get_owner',
                            type: 'function',
                            inputs: [],
                            outputs: [
                                {
                                    type: 'core::felt252',
                                },
                            ],
                            state_mutability: 'view',
                        },
                    ],
                },
            ],
            selector: 'get_owner',
            version: '1.0.0',
            postProcessor: hexStringPostProcessor,
        },
    ],
    [
        2,
        {
            name: 'Braavos',
            abi: [
                {
                    name: 'braavos_account::signers::interface::ISignerManagement',
                    type: 'interface',
                    items: [
                        {
                            name: 'get_public_key',
                            type: 'function',
                            inputs: [],
                            outputs: [
                                {
                                    type: 'core::felt252',
                                },
                            ],
                            state_mutability: 'view',
                        },
                    ],
                },
            ],
            selector: 'get_public_key',
            version: '1.0.0',
            postProcessor: hexStringPostProcessor,
        },
    ],
]);
